Nightly Serial build failure, TestStringManipulation.cpp: error: static assertion failed

Nightly Serial build of Trilinos with Kokkos' develop branch failed to compile with intel/18 with the following message

```
[ 23%] Building CXX object kokkos/core/unit_test/CMakeFiles/KokkosCore_UnitTest_KokkosP.dir/UnitTestMainInit.cpp.o
<https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Intel181163_Serial_Cplx/ws/Trilinos/kokkos/core/unit_test/TestStringManipulation.cpp(71)>: error: static assertion failed with ""
    STATIC_ASSERT(strcmp(cat1, cat2) < 0);
```

I'll update the issue as more testing completes to give a better scope of which builds are affected

Also failing with older gcc, at least with gcc/5.3.0

```
18:41:36 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkos-dev_gcc_530_deprecated_off/kokkos/core/unit_test/TestStringManipulation.cpp:50:29: error: non-constant condition for static assertion
18:41:36  #define STATIC_ASSERT(cond) static_assert(cond, "")
```

--- a/core/unit_test/TestStringManipulation.cpp
+++ b/core/unit_test/TestStringManipulation.cpp
@@ -66,8 +66,10 @@ KOKKOS_FUNCTION constexpr bool test_strcmp() {
   constexpr char cat3[] = "Hobbes";
   constexpr char cat4[] = "Garfield";
   STATIC_ASSERT(strcmp(cat1, cat1) == 0);
-#if !defined(KOKKOS_COMPILER_NVCC) || \
-    ((__CUDACC_VER_MAJOR__ >= 11) && (__CUDACC_VER_MINOR__ >= 3))
+#if (!defined(KOKKOS_COMPILER_NVCC) ||                                 \
+     ((__CUDACC_VER_MAJOR__ >= 11) && (__CUDACC_VER_MINOR__ >= 3))) && \
+    (!defined(__INTEL_COMPILER_BUILD_DATE) ||                          \
+     (__INTEL_COMPILER_BUILD_DATE >= 20210228))
   STATIC_ASSERT(strcmp(cat1, cat2) < 0);
   STATIC_ASSERT(strcmp(cat1, cat3) < 0);
 #endif

--- a/core/unit_test/TestStringManipulation.cpp
+++ b/core/unit_test/TestStringManipulation.cpp
@@ -178,6 +178,7 @@ KOKKOS_FUNCTION constexpr bool test_strncat() {
 }
 STATIC_ASSERT(test_strncat());
 
+#if !defined(KOKKOS_COMPILER_GNU) && (KOKKOS_COMPILER_GNU >= 540)
 template <class Integral>
 KOKKOS_FUNCTION constexpr bool to_chars_helper(Integral val, char const* ref) {
   using Kokkos::Impl::strcmp;
@@ -211,5 +212,6 @@ KOKKOS_FUNCTION constexpr bool test_to_chars() {
   return true;
 }
 STATIC_ASSERT(test_to_chars());
+#endif
 
 }  // namespace

--- a/core/unit_test/TestStringManipulation.cpp
+++ b/core/unit_test/TestStringManipulation.cpp
@@ -95,7 +95,7 @@ KOKKOS_FUNCTION constexpr bool test_strncmp() {
 #if defined(KOKKOS_COMPILER_GNU) && (KOKKOS_COMPILER_GNU < 610)
   (void)greet3;
 #elif defined(KOKKOS_COMPILER_GNU) && (KOKKOS_COMPILER_GNU < 710)
-  STATIC_ASSERT(strncmp(&greet2[12], &greet3[11] + 11, 5) == 0);
+  STATIC_ASSERT(strncmp(&greet2[12], &greet3[11], 5) == 0);
 #else
   STATIC_ASSERT(strncmp(greet2 + 12, greet3 + 11, 5) == 0);
 #endif

